src/Components/Components/src/Routing/RouteTableFactory.cs

```diff
@@ -56,7 +56,10 @@ static void GetRouteableComponents(List<Type> routeableComponents, Assembly asse
             {
                 if (typeof(IComponent).IsAssignableFrom(type) && type.IsDefined(typeof(RouteAttribute)))
                 {
-                    routeableComponents.Add(type);
+                    if (!routeableComponents.Contains(type))
```

Comment:
This is going to do a linear scan isn't it? Could you tell me what the error was when you used a HashSet instead? 

An alternative fix is to de-dupe AppAssembly and AdditionalAssemblies. Essentially you could check on line 44 if `assembly != routeKey.AppAssembly` before calling `GetRouteableComponents`

Reply:
> Could you tell me what the error was when you used a HashSet instead?

The tests are failing with the following reason ([example in Azure DevOps](https://dev.azure.com/dnceng/public/_build/results?buildId=1521937&view=logs&j=5ac7b393-e840-5549-7fb4-a4479af8e7e3&t=29df2fa2-0d20-51bd-e85a-8b546e86c529&l=1506)):

> Using member 'System.Reflection.Assembly.ExportedTypes.get' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Types might be removed.

I just replaced the `List` with a `HashSet` and adjusted other references in the [commit](https://github.com/dotnet/aspnetcore/pull/39129/commits/bbb9dbe1ef7085e2e820b4d4d51045398749ed5f), but it was failing.

> An alternative fix is to de-dupe AppAssembly and AdditionalAssemblies.

This would also be possible, will change the PR to that solution.